### ISSUE TITLE
`volnmf_main` function is broken

### ISSUE THREAD

**Issue by VPetukhov** [kharchenkolab/vrnmf#4]
I don't know when it happened, but now following [volume_regularized_NMF.md](https://github.com/kharchenkolab/vrnmf/blob/main/doc/volume_regularized_NMF.md) causes an error during the call to `volnmf_main`:
> Error in matrix(runif(nrow(B) * n.comp, 0, 1), nrow = nrow(B), ncol = n.comp) : 
  non-numeric matrix extent

And looking at the code, there is a weird `mutation.run` parameter that is `FALSE` by default, which [sets](https://github.com/kharchenkolab/vrnmf/blob/176ca8f139c1bb57a9518a70473d353c562de970/R/volnmf_estimate.R#L72) `vol <- NULL`. Twice, btw (also [here](https://github.com/kharchenkolab/vrnmf/blob/176ca8f139c1bb57a9518a70473d353c562de970/R/volnmf_estimate.R#L78)). Given that **all** information is stored in `vol` and it's removed as the very first thing in the function, the rest doesn't work. Particularly, it fails now as `B` [can't be estimated](https://github.com/kharchenkolab/vrnmf/blob/176ca8f139c1bb57a9518a70473d353c562de970/R/volnmf_estimate.R#L83), and so the code [tries to get](https://github.com/kharchenkolab/vrnmf/blob/176ca8f139c1bb57a9518a70473d353c562de970/R/volnmf_estimate.R#L110) `nrow(NULL)` and fails.

@solrust , I think only you really know what happens there.

**Comment by evanbiederstedt** [comment]
FWIW, I tried installing @solrust original code here: https://github.com/kharchenkolab/vrnmf/releases/tag/v0.1.0

I get the same error. 
```
> volres <- volnmf_main(vol, n.comp = 10, wvol = 2e-2, 
                     n.iter = 3e+3, vol.iter = 1e+1, c.iter = 1e+1, 
                    extrapolate = TRUE, accelerate = TRUE, verbose = FALSE)
Error in matrix(runif(nrow(B) * n.comp, 0, 1), nrow = nrow(B), ncol = n.comp) : 
  non-numeric matrix extent
```

But the Rmarkdown files rendered back then: https://github.com/kharchenkolab/vrnmf/blob/main/doc/volume_regularized_NMF.md

Maybe there's been a change with the dependencies?

What I find confusing is that this is an error within base R, not a package:

```
> matrix(1, nrow = "foo", ncol = 1)
Error in matrix(1, nrow = "foo", ncol = 1) : non-numeric matrix extent
```

I agree---it would be best if @solrust tries debugging this first.

**Comment by andrewdchen** [comment]
Interestingly, it doesn't seem like resetting to a previous commit (before [74b6c91](https://github.com/kharchenkolab/vrnmf/commit/74b6c91f0a6bbaea95bffad3ca002518eb19c872)) where `vol <- NULL` doesn't fix the issue.

**Comment by evanbiederstedt** [comment]
I had a chance to look into this. (Thanks for the interest @andrewdchen! )

Andrew is on the right track. The error is from here: https://github.com/kharchenkolab/vrnmf/commit/89b8ab1c0a51376de2a01423bd52d7d7d5d6f06e

**Code:**
```
library(vrnmf)

simnmf <- simulatedNMF::simnmf

vol <- vol_preprocess(t(simnmf$X))

volres <- volnmf_main(vol, n.comp = 10, wvol = 2e-2, 
                    n.iter = 3e+3, vol.iter = 1e+1, c.iter = 1e+1, 
                    extrapolate = TRUE, accelerate = TRUE, verbose = FALSE)
```

**Error**
```
Error in matrix(numeric(0), nrow = NULL, ncol = 10) : 
  non-numeric matrix extent
```

**Explanation for buggy behavior:**

By default, `mutation.run = FALSE`. 


First line is this:

```
  if (!mutation.run){
    rate.rec <- xcompl <- vol <- NULL
  }
```

That sets `rect.rec`, `xcmopl`, and `vol` to `NULL`

I'm not sure what @solrust wanted to do here....it doesn't quite make sense to reset `vol` to `NULL` after instantiating this variable above. 

But that logic leads to bugs here:

```
  # matrix 
  if (is.null(B)){
    if (!is.null(volnmf)){
      B <- volnmf$B
    }else if(!is.null(vol)){
      B <- -vol$U[,1:n.reduce]%*%sqrt(diag(vol$eigens)[1:n.reduce,1:n.reduce])
    }
  }
```

The parameter `B` is set to `NULL` by default. But now, `!is.null(volnmf)` is `FALSE`, and `!is.null(vol)` is `FALSE`; none of those conditionals are met. 

Previously, this would work:

```
}else if(!is.null(vol)){
  B <- -vol$U[,1:n.reduce]%*%sqrt(diag(vol$eigens)[1:n.reduce,1:n.reduce])
}
```

but currently `B` remains `NULL`. 

Given `!is.null(anchor)` is also `NULL` and`B` remains `NULL` , the next code run is here:


` C <- matrix(runif(nrow(B)*n.comp, 0, 1), nrow = nrow(B), ncol = n.comp)`

That's the bug we see. 

It's actually running `matrix(numeric(0), nrow = NULL, ncol = 10)`...

```
> matrix(numeric(0), nrow = NULL, ncol = 10)
Error in matrix(numeric(0), nrow = NULL, ncol = 10) : 
  non-numeric matrix extent
```

When Ruslan clarifies things, we can fix this. CC @VPetukhov 



**PS:** On a separate issue, I don't know how this happened, but it looks like my fault. This is duplicated code; possible a merge mistake. We should fix this too:

```
  if (mutation.run==FALSE) {
    rate.rec <- NULL
    xcompl <- NULL
    vol <- NULL
  }

  if (!mutation.run){
    rate.rec <- xcompl <- vol <- NULL
  }
```

**Comment by andrewdchen** [comment]
Gotcha, that makes a lot of sense! It does seem to be the changes in that commit that are problematic - resetting to [6e1cae7](https://github.com/kharchenkolab/vrnmf/commit/6e1cae77ea3f7d216f569c482693abc9704d1047), parent of [89b8ab1](https://github.com/kharchenkolab/vrnmf/commit/89b8ab1c0a51376de2a01423bd52d7d7d5d6f06e) yields a working `volnmf_main`

**Comment by evanbiederstedt** [comment]
Should be fixed with v1.0.2; going on CRAN shortly
